### PR TITLE
Decompile sub_6CA2DF

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/windows/ride_construction.c [hunk 1]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+	if (_enabledRidePiecesB & (1 << 8)) {
+		if (trackType == 0x3E) {
+			trackType = 0x76;
+		}
+
+		if (trackType == 0x3F) {
+			trackType = 0x77;
+		}
+
+		if (trackType == 0x48) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: 0x40

File: src/windows/ride_construction.c [hunk 2]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+			trackType = 0x7A;
+		}
+
+		if (trackType == 14) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: 0x41

File: src/windows/ride_construction.c [hunk 3]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+			trackType = 0x79;
+		}
+
+		if (trackType == 0x9A) {
+			return true;
+		}
+
+		if (trackType == 0x98) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: 0x9B

File: src/windows/ride_construction.c [hunk 4]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+		if (trackType == 0x9D) {
+			return true;
+		}
+
+		if (trackType == 0x9C) {
+			return true;
+		}
+	}
+
+	if (ride_type_has_flag(ride->type, 0x1000) && _currentTrackCovered != 1) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: `_currentTrackCovered & 1`

File: src/windows/ride_construction.c [hunk 5]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+		if (ride->type == RIDE_TYPE_WATER_COASTER || trackType == 0 || trackType == 0x10 || trackType == 0x11) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: `!= RIDE_TYPE_WATER_COASTER`

File: src/windows/ride_construction.c [hunk 6]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+			trackType = RCT2_GLOBAL(0x00993D1C + trackType, uint8);
+			edxRS16 &= 0xFFFE; // unsets 0x1
+		}
+	}
+
+	z = _currentTrackBeginZ;
+	rct_track_coordinates trackCoordinates;
+	if (ride_type_has_flag(ride->type, 0x80000)) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: There is a function for getting the correct `TrackCoordinates`.

File: src/windows/ride_construction.c [hunk 7]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+	x = _currentTrackBeginX;
+	y = _currentTrackBeginY;
+	if (_rideConstructionState == 2) {
+		z -= trackCoordinates.z_end;
+		trackDirection = RCT2_GLOBAL(RCT2_ADDRESS_TRACK_PREVIEW_ROTATION, uint8) ^ 0x02;
+		trackDirection -= trackCoordinates.rotation_end;
+		trackDirection += trackCoordinates.rotation_begin;
+		trackDirection &= 0x03;
+
+		if (trackCoordinates.rotation_begin == 4) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: & (1<<2)

File: src/windows/ride_construction.c [hunk 8]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+	if (_rideConstructionState == RIDE_CONSTRUCTION_STATE_BACK) {
+		bl = _currentTrackSlopeEnd;
+		bh = _previousTrackSlopeEnd;
+		cl = _currentTrackBankEnd;
+		ch = _previousTrackBankEnd;
+	}
+
+	uint16 ax = _currentTrackCurve;
+	if (ax == 0xFFFF) {
+		return true;
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: return false?

File: src/windows/ride_construction.c [hunk 9]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+						*dh = 0x12;
+						if (ch == 2) {
+							return true;
+						}
+
+						*dh = 0x13;
+						return true;
+					}
+
+					if (ch == 0) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: ch != 0 return false

File: src/windows/ride_construction.c [hunk 10]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+						*dh = 0xE1;
+						if (ch == 2) {
+							return true;
+						}
+
+						*dh = 0xE2;
+						return true;
+					}
+
+					if (ch == 0) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: `ch != 0`

File: src/windows/ride_construction.c [hunk 11]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+
+				if (bl == 6) {
+					if (cl == ch) {
+						*dh = 0x73;
+						if (cl == 2) {
+							return true;
+						}
+
+						*dh = 0x74;
+						if (cl == 2) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: cl == 4

File: src/windows/ride_construction.c [hunk 12]
```diff
@@ -2380,6 +2381,1292 @@ void sub_6C84CE()
...
+						*dh = 0xE5;
+						if (ch == 2) {
+							return true;
+						}
+
+						*dh = 0xE6;
+						return true;
+					}
+
+					if (ch == 0) {
```
Comment (duncanspumpkin) [inline src/windows/ride_construction.c]: `!= 0`

**Comment by marijnvdwerf** [comment]
Some other combinations which I haven't been able to trigger so far:

```c
// Fully known, but not yet ecountered
{	false,	TRACK_SLOPE_DOWN_25,	TRACK_BANK_NONE,	TRACK_CURVE_LEFT_SMALL,			TRACK_SLOPE_NONE,		TRACK_BANK_LEFT,	TRACK_ELEM_LEFT_QUARTER_TURN_3_TILES_25_DEG_DOWN_TO_LEFT_BANK		},
{	false,	TRACK_SLOPE_DOWN_25,	TRACK_BANK_NONE,	TRACK_CURVE_RIGHT_SMALL,		TRACK_SLOPE_NONE,		TRACK_BANK_RIGHT,	TRACK_ELEM_LEFT_QUARTER_TURN_3_TILES_25_DEG_DOWN_TO_LEFT_BANK		},

// Partially known
{	false,	TRACK_SLOPE_DOWN_90,	0,					TRACK_CURVE_LEFT_VERY_SMALL,	TRACK_SLOPE_DOWN_90,	0,					TRACK_ELEM_LEFT_QUARTER_TURN_1_TILE_90_DEG_DOWN						},
{	false,	TRACK_SLOPE_DOWN_90,	0,					TRACK_CURVE_RIGHT_VERY_SMALL,	TRACK_SLOPE_DOWN_90,	0,					TRACK_ELEM_RIGHT_QUARTER_TURN_1_TILE_90_DEG_DOWN					},
{	false,	TRACK_SLOPE_DOWN_60,	0,					TRACK_CURVE_LEFT_VERY_SMALL,	TRACK_SLOPE_DOWN_60,	0,					TRACK_ELEM_LEFT_QUARTER_TURN_1_TILE_60_DEG_DOWN						},
{	false,	TRACK_SLOPE_DOWN_60,	0,					TRACK_CURVE_RIGHT_VERY_SMALL,	TRACK_SLOPE_DOWN_60,	0,					TRACK_ELEM_RIGHT_QUARTER_TURN_1_TILE_60_DEG_DOWN					},
{	false,	TRACK_SLOPE_UP_60,		0,					TRACK_CURVE_LEFT_VERY_SMALL,	TRACK_SLOPE_UP_60,		0,					TRACK_ELEM_LEFT_QUARTER_TURN_1_TILE_60_DEG_UP						},
{	false,	TRACK_SLOPE_UP_60,		0,					TRACK_CURVE_RIGHT_VERY_SMALL,	TRACK_SLOPE_UP_60,		0,					TRACK_ELEM_RIGHT_QUARTER_TURN_1_TILE_60_DEG_UP						},
{	false,	TRACK_SLOPE_UP_90,		0,					TRACK_CURVE_LEFT_VERY_SMALL,	TRACK_SLOPE_UP_90,		0,					TRACK_ELEM_LEFT_QUARTER_TURN_1_TILE_90_DEG_UP						},
{	false,	TRACK_SLOPE_UP_90,		0,					TRACK_CURVE_RIGHT_VERY_SMALL,	TRACK_SLOPE_UP_90,		0,					TRACK_ELEM_RIGHT_QUARTER_TURN_1_TILE_90_DEG_UP						},
```

**Pull request merged**